### PR TITLE
C++: Remove more field-to-object flow

### PR DESCRIPTION
**Pull request by MathiasVP**
Copy/paste from my comment [here](https://github.com/github/codeql-c-analysis-team/issues/103):

-----------
While looking through the CPP-differences changes for https://github.com/github/codeql/pull/3364 I found another source of field -> object taint in `simpleInstructionLocalFlowStep` in `DataFlowUtil`:

```codeql
private predicate simpleInstructionLocalFlowStep(Operand opFrom, Instruction iTo) {
  // ...
  exists(ChiInstruction chi | chi = iTo |
      opFrom.getAnyDef() instanceof WriteSideEffectInstruction and
      chi.getPartialOperand() = opFrom and
      not chi.isResultConflated()
    )
  // ..
}
```

This rule ensures that we get flow from the `WriteSideEffect` after calling `set_value` to the argument to `sink` in this example:
```cpp
void set_value(int*);

void test() {
  int n;
  set_value(&n);
  sink(n);
}
```
however, it also transfers flow from the `WriteSideEffect` to the `sink` argument in the following case:
```cpp
void test() {
  struct { int field1, field2; } obj;
  set_value(&obj->field1);
  sink(obj);
}
```
this pattern gave rise to some field conflation on `cpp/uncontrolled-allocation-size` on OpenJDK and on `cpp/unbounded-write` on vim.

The fix is to restrict the rule in `simpleInstructionLocalFlowStep` to only propagate flow from the `WriteSideEffect` to the `Chi` when we know that the `Chi` isn't only a partial update. <s>We should be able to use the `getUpdatedInterval` predicate on the `ChiInstruction` for this.</s>

-----------
I was right about everything except for the very last line. It's not so easy to use `getUpdatedInterval` to detect whether the entire memory location is updated since (AFAIK) it's not possible to get the type of the memory location without digging into SSA. So I added a new predicate on `ChiInstruction` that forwards to a new predicate in `SSAConstruction.qll` which then does the actual work.

Unfortunately, we lose a couple of results in dataflow. This is due to field flow not looking "deep enough" through the destination address on write side effects (i.e., we only look check whether the destination instruction is a `FieldAddress` or not, but we should traverse through the entire chain of instructions like in https://github.com/github/codeql/pull/5127). So `simpleInstructionLocalFlowStep` expects field flow to handle this case, but it currently doesn't.

@dbartol / @rdmarsh2  I'd appreciate your thoughts on the first commit (i.e. 200d94777adc5e3fa024cddd52042772819619cc).

(Part of https://github.com/github/codeql-c-analysis-team/issues/103.)

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/internal/SSAConstruction.qll [hunk 1]
```diff
@@ -178,6 +178,22 @@ private module Cached {
     )
   }
 
+  /**
+   * Holds if the `ChiPartialOperand` totally, but not exactly, overlaps with the `ChiTotalOperand`.
```
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/internal/SSAConstruction.qll:182]: This should be "partially overlaps with", I believe. A "total" overlap overwrites all bits but could be a larger write or with a different type, an "exact" overlap writes with the same type and overwrites all bits.

**Review by rdmarsh2 — COMMENTED** [review]
Do we have a test case where this improves the results, or is it only to make #3364 work correctly?

**Comment by MathiasVP** [comment]
> Do we have a test case where this improves the results, or is it only to make #3364 work correctly?

Fixed in bd842403c846589f52e8508f4b807aa2af1e2ede. I'm surprised the flow to `sink(s.x)` isn't caught, but at least it's not a regression.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/internal/SSAConstruction.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/internal/SSAConstruction.qll]: Good point. I've fixed this in c86fc223b938129421f2e9a5535b7d3ee87ffd40.

**Review by jbj — APPROVED** [review]
LGTM. I'll merge this to unblock further work. @dbartol, do come back with a review if you have any reservations -- there is still time to change the interface or the implementation.

**Comment by MathiasVP** [comment]
I forgot to link to the CPP-differences results. [Here they are](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1789/) in case we need to look at them again later.

**Pull request merged**